src/Backups/BackupsWorker.cpp

```diff
@@ -218,42 +218,145 @@ namespace
+            /// `use_queue == true` - put into the thread pool's queue,
+            /// `use_queue == false` - schedule() should wait until some of the jobs finish.
+            bool use_queue = false;
+
+            switch (thread_pool_id)
+            {
+                case ThreadPoolId::BACKUP_ASYNC:
+                case ThreadPoolId::BACKUP_ASYNC_ON_CLUSTER:
+                case ThreadPoolId::BACKUP_COPY_FILES:
+                {
+                    metric_threads = CurrentMetrics::BackupsThreads;
+                    metric_active_threads = CurrentMetrics::BackupsThreadsActive;
+                    max_threads = num_backup_threads;
+                    /// We don't use the thread pool's queue for copying files because otherwise that queue could be memory-wasting.
+                    use_queue = (thread_pool_id != ThreadPoolId::BACKUP_COPY_FILES);
+                    break;
+                }
+
+                case ThreadPoolId::RESTORE_ASYNC:
+                case ThreadPoolId::RESTORE_ASYNC_ON_CLUSTER:
+                case ThreadPoolId::RESTORE_TABLES_DATA:
+                {
+                    metric_threads = CurrentMetrics::RestoreThreads;
+                    metric_active_threads = CurrentMetrics::RestoreThreadsActive;
+                    max_threads = num_restore_threads;
+                    use_queue = (thread_pool_id != ThreadPoolId::RESTORE_TABLES_DATA);
+                    break;
+                }
+
+                default:
+                    UNREACHABLE();
+            }
+
+            /// We set max_free_threads = 0 because we don't want to keep any threads if there is no RESTORE query running right now.
+            size_t max_free_threads = 0;
+            size_t queue_size = use_queue ? 0 : max_threads;
+            thread_pool = std::make_unique<ThreadPool>(metric_threads, metric_active_threads, max_threads, max_free_threads, queue_size);
+        }
+        return *thread_pool;
+    }
+
+    /// Waits for all threads to finish.
+    void wait()
+    {
+        auto wait_sequence = {
+            ThreadPoolId::RESTORE_ASYNC_ON_CLUSTER,
+            ThreadPoolId::RESTORE_ASYNC,
+            ThreadPoolId::RESTORE_TABLES_DATA,
+            ThreadPoolId::BACKUP_ASYNC_ON_CLUSTER,
+            ThreadPoolId::BACKUP_ASYNC,
+            ThreadPoolId::BACKUP_COPY_FILES,
+        };
+
+        for (auto thread_pool_id : wait_sequence)
+        {
+            ThreadPool * thread_pool;
+            {
+                std::lock_guard lock{mutex};
+                thread_pool = thread_pools[static_cast<size_t>(thread_pool_id)].get();
+            }
+            if (thread_pool)
+                thread_pool->wait();
+        }
+    }
+
+private:
+    const size_t num_backup_threads;
+    const size_t num_restore_threads;
+    std::unique_ptr<ThreadPool> thread_pools[static_cast<size_t>(ThreadPoolId::MAX)] TSA_GUARDED_BY(mutex);
```

Comment:
Maybe better to make unordered map for simplicity?

Reply:
done